### DISCUSSION TITLE
Library for Go

### DISCUSSION THREAD

**Discussion by Agwave** [apache/incubator-graphar#540] [General]
Does this project have plans to provide a library for Go language？

**Comment by acezen** [comment]
Hi, @Agwave, thanks for bring that up. We don't have plan to provide Go library now. We do want to provide more language libraries, but don't have much manpower to do that. If you have interest, we are very appreciated you can work on it.

**Reply by SemyonSinchenko** [reply]
I think we can setup automated build of the go classes from the protobuf messages, it is just one more line in `buf.gen.yml`; But not a full featured API.

**Reply by Agwave** [reply]
Thank you for your reply, I understand that there are no plans to provide a Go language library at the moment, and I am indeed hesitating. However, if I have sufficient skills and time, I would be very willing to participate in this project. Thank you again for this opportunity, and I look forward to seeing the future development of the project.

**Reply by acezen** [reply]
> I think we can setup automated build of the go classes from the protobuf messages, it is just one more line in `buf.gen.yml`; But not a full featured API.

yes, we can provide basic info implementation by generated the protobuf with go.

**Reply by xin-hao-2025** [reply]
maybe we can add the updates or plans here https://github.com/apache/incubator-graphar/issues/72
personally I'm caring about the Rust library, and I think if we have Rust library, we will of course have the Python library (by [PyO3](https://github.com/PyO3/pyo3)).

**Reply by SemyonSinchenko** [reply]
> and I think if we have Rust library, we will of course have the Python library (by [PyO3](https://github.com/PyO3/pyo3)).

But we can do the same with ffi and the existing cpp library.

**Reply by acezen** [reply]
> maybe we can add the updates or plans here #72 personally I'm caring about the Rust library, and I think if we have Rust library, we will of course have the Python library (by [PyO3](https://github.com/PyO3/pyo3)).

Thanks for the feedback, support more libraries for GraphAr is one of our goals for the project. But we need to consider a good strategy to implement them which need not much manpower to maintain them. Could you describe the use case or scenario you wan to use GraphAr? like what graph database you want to integrate? That would be very helpful to our support to the library.

**Reply by xin-hao-2025** [reply]
> like what graph database you want to integrate?

[Nebula](https://github.com/vesoft-inc/nebula)

**Reply by acezen** [reply]
> > like what graph database you want to integrate?
> 
> [Nebula](https://github.com/vesoft-inc/nebula)

thanks for the reply, but I found that nebula is base on C++, is there anything I miss that it has the rust module?

**Reply by xin-hao-2025** [reply]
> but I found that nebula is base on C++

you are right, I'm planning for the future Nebula v5 and that is my personal preference.
sry for the confusing

**Reply by acezen** [reply]
> > but I found that nebula is base on C++
> 
> you are right, I'm planning for the future Nebula v5 and that is my personal preference. sry for the confusing

Thanks for giving the message,  it would be driving force for us to provide the rust library:)

**Reply by ZekiLiu** [reply]
> Hi, @Agwave, thanks for bring that up. We don't have plan to provide Go library now. We do want to provide more language libraries, but don't have much manpower to do that. If you have interest, we are very appreciated you can work on it.
> 
> 
> 嗨，@Agwave，谢谢你把它拿出来。我们现在没有提供Go库的计划。我们确实想提供更多的语言库，但没有太多的人力来做这件事。如果你有兴趣，我们非常感谢你可以在它的工作。

Hi! I have drafted a detailed implementation plan for the native Go SDK. I propose a phased approach starting with a Read-Only implementation.

Please check the full proposal here: [#828](https://github.com/apache/incubator-graphar/issues/828)

I'm ready to start Phase 1 once we align on the design
